File: NodejsConsoleApp1/simulate_fight_2.js
// Roll a die with a specified number of sides
function roll_d(size) {
    return Math.floor(Math.random() * size) + 1;
}

// Extract damage details from a damage string
function get_damage(damage) {
    const damage_parts = damage.damage_dice.split('+');
    const dice_parts = damage_parts[0].split('d');
    const num_dice = parseInt(dice_parts[0]);
    const die_type = parseInt(dice_parts[1]);
    const damage_bonus = parseInt(damage_parts[1]) || 0;
    return [num_dice, die_type, damage_bonus];
}

// Fetch creature stats from the D&D 5e API
async function get_creature_stats(creature_name) {
    const response = await fetch(`https://www.dnd5eapi.co/api/monsters/${creature_name.toLowerCase().replace(' ', '-')}`);

    if (response.status === 200) {
        const data = await response.json();

        // Extract relevant stats
        const initiative_bonus = Math.floor((data.dexterity - 10) / 2);

        // Determine attacks from multiattack if available, otherwise use the first action
        const multiattack = data.actions.find((action) => action.name.toLowerCase() === 'multiattack');
        let attacks = [];

        if (multiattack && multiattack.actions) {
            // Collect actions specified in multiattack
            for (const action_info of multiattack.actions) {
                const action_name = action_info.action_name;
                const count = action_info.count;
                // Find the matching action details
                const action = data.actions.find((act) => act.name === action_name);

                if (action && action.damage) {
                    const attack_details = {
                        count: count,
                        attack_bonus: action.attack_bonus,
                        damage_sets: [],
                    };

                    for (const damage of action.damage) {
                        attack_details.damage_sets.push(get_damage(damage));
                    }

                    attacks.push(attack_details);
                }
            }
        } else {
            // Fall back to using the first action if no multiattack is present
            if (data.actions) {
                const action = data.actions[0];

                if (action.damage) {
                    const attack_details = {
                        count: 1,
                        attack_bonus: action.attack_bonus,
                        damage_sets: [],
                    };

                    for (const damage of action.damage) {
                        attack_details.damage_sets.push(get_damage(damage));
                    }

                    attacks.push(attack_details);
                }
            }
        }

        return {
            name: data.name,
            initiative_bonus: initiative_bonus,
            hp: data.hit_points, 
            ac: data.armor_class[0].value,
            attacks: attacks
        };
    } else {
        throw new Error(`Could not retrieve stats for ${creature_name}`);
    }
}

// Simulate a fight between two D&D 5e creatures
async function simulate_fight(c1, c2) {
    // Roll initiative
    const c1_init = roll_d(20) + c1.initiative_bonus;
    const c2_init = roll_d(20) + c2.initiative_bonus;

    // Determine attacker and defender based on initiative
    let attacker, attacker_hp, defender, defender_hp;
    if (c1_init >= c2_init) {
        [attacker, attacker_hp, defender, defender_hp] = [c1, c1.hp, c2, c2.hp];
    } else {
        [attacker, attacker_hp, defender, defender_hp] = [c2, c2.hp, c1, c1.hp];
    }

    while (attacker_hp > 0) {
        // Perform each attack in the attacker's attacks list
        for (const attack of attacker.attacks) {
            for (let i = 0; i < attack.count; i++) {
                // Roll to hit
                const attack_roll = roll_d(20) + attack.attack_bonus;

                // If hit, calculate damage
                if (attack_roll >= defender.ac) {
                    let total_damage = 0;

                    for (const [num_dice, die_type, damage_bonus] of attack.damage_sets) {
                        let damage = damage_bonus; // Start with damage bonus

                        for (let j = 0; j < num_dice; j++) {
                            damage += roll_d(die_type); // Type of damage die
                        }

                        if (attack_roll === 20 + attack.attack_bonus) // Critical hit
                            damage += damage;

                        total_damage += damage;
                    }

                    defender_hp -= total_damage;
                }
            }

            // Switch attacker and defender
            [attacker, attacker_hp, defender, defender_hp] = [defender, defender_hp, attacker, attacker_hp];
        }
    }

    return defender.name;
}

// Run a specified number of fights and return the win percentages
async function run_simulations(c1_name, c2_name, num_fights) {
    const c1 = await get_creature_stats(c1_name);
    const c2 = await get_creature_stats(c2_name);

    let c1_wins = 0;
    let c2_wins = 0;

    for (let i = 0; i < num_fights; i++) {
        const winner = await simulate_fight(c1, c2);

        if (winner === c1_name) {
            c1_wins++;
        } else {
            c2_wins++;
        }
    //    console.clear();
    //    console.log(`${c1.name} wins: ${c1_wins}; ${c2.name} wins: ${c2_wins};`);
    }

    const c1_win_percentage = (c1_wins / num_fights) * 100;
    const c2_win_percentage = (c2_wins / num_fights) * 100;

    return [c1_win_percentage, c2_win_percentage];
}

// Run 1000 simulations
const c1_name = 'Owlbear';
const c2_name = 'Unicorn';
const [c1_win_percentage, c2_win_percentage] = await run_simulations(c1_name, c2_name, 1000);

console.log(`${c1} win percentage: ${c1_win_percentage.toFixed(2)}%`);
console.log(`${c2} win percentage: ${c2_win_percentage.toFixed(2)}%`);

File: ForFunsies/simulate_fight_2.py
import random
import requests
import os
clear = lambda: os.system('cls')

def roll_d(size):
    return random.randint(1, size)

def get_damage(damage):
    damage_parts = damage['damage_dice'].split('+')
    dice_parts = damage_parts[0].split('d')
    num_dice = int(dice_parts[0])
    die_type = int(dice_parts[1])
    damage_bonus = int(damage_parts[1]) or 0
    return (num_dice, die_type, damage_bonus)

def get_creature_stats(creature_name):
    """Fetches the creature's stats from the D&D 5e API."""
    response = requests.get(f"https://www.dnd5eapi.co/api/monsters/{creature_name.lower().replace(' ', '-')}")
    if response.status_code == 200:
        data = response.json()

        # Extract relevant stats
        name = data['name']
        dexterity = data['dexterity']
        initiative_bonus = (dexterity - 10) // 2  # Calculate initiative bonus from dexterity
        hp = data['hit_points']
        ac = data['armor_class'][0]['value']

        # Determine attacks from multiattack if available, otherwise use the first action
        multiattack = next((action for action in data['actions'] if action['name'].lower() == 'multiattack'), None)
        attacks = []

        if multiattack and 'actions' in multiattack:
            # Collect actions specified in multiattack
            for action_info in multiattack['actions']:
                action_name = action_info['action_name']
                count = action_info['count']
                # Find the matching action details
                action = next((act for act in data['actions'] if act['name'] == action_name), None)
                if action and 'damage' in action:
                    attack_details = {
                        'count': count,
                        'attack_bonus': action['attack_bonus'],
                        'damage_sets': []
                    }
                    for damage in action['damage']:
                        attack_details['damage_sets'].append(get_damage(damage))
                    attacks.append(attack_details)
        else:
            # Fall back to using the first action if no multiattack is present
            if data['actions']:
                action = data['actions'][0]
                if 'damage' in action:
                    attack_details = {
                        'count': 1,
                        'attack_bonus': action['attack_bonus'],
                        'damage_sets': []
                    }
                    for damage in action['damage']:
                        attack_details['damage_sets'].append(get_damage(damage))
                    attacks.append(attack_details)

        return name, initiative_bonus, hp, ac, attacks
    else:
        raise ValueError(f"Could not retrieve stats for {creature_name}")

def simulate_fight(c1, c2):
    """Simulates a basic fight between two D&D 5e creatures and returns the winner."""
    # Roll initiative
    c1_init = roll_d(20) + c1[1]
    c2_init = roll_d(20) + c2[1]

    # Determine attacker and defender based on initiative
    if c1_init >= c2_init:
        attacker, attacker_hp, defender, defender_hp = c1, c1[2], c2, c2[2]
    else:
        attacker, attacker_hp, defender, defender_hp = c2, c2[2], c1, c1[2]

    while attacker_hp > 0:
        # Perform each attack in the attacker's attacks list
        for attack in attacker[4]:
            for _ in range(attack['count']):  # Number of attacks
                attack_roll = roll_d(20) + attack['attack_bonus']  # Bonus to hit
                if attack_roll >= defender[3]:  # Defender's AC
                    total_damage = 0
                    for num_dice, die_type, damage_bonus in attack['damage_sets']:
                        damage = damage_bonus  # Start with damage bonus
                        if attack_roll == 20 + attack['attack_bonus']:  # If it was a critical hit
                            num_dice *= 2  # Double the number of damage dice
                        for _ in range(num_dice):
                            damage += roll_d(die_type)  # Type of damage die
                        total_damage += damage
                    defender_hp -= total_damage

        temp, temp_hp = attacker, attacker_hp
        attacker, attacker_hp = defender, defender_hp
        defender, defender_hp = temp, temp_hp

    return defender[0]

def run_simulations(c1_name, c2_name, num_fights):
    """Runs a specified number of fights and returns the win percentages."""
    c1 = get_creature_stats(c1_name)
    c2 = get_creature_stats(c2_name)


    c1_wins = 0
    c2_wins = 0

    for _ in range(num_fights):
        winner = simulate_fight(c1, c2)
        if winner == c1_name:
            c1_wins += 1
        else:
            c2_wins += 1
        # clear()
        # print(f"{c1_name} wins: {c1_wins}; {c2_name} wins: {c2_wins};")

    c1_win_percentage = (c1_wins / num_fights) * 100
    c2_win_percentage = (c2_wins / num_fights) * 100


    return c1_win_percentage, c2_win_percentage

# Run 100000 simulations
c1 = "Owlbear"
c2 = "Unicorn"
c1_win_percentage, c2_win_percentage = run_simulations(c1, c2, 100000)

print(f"{c1} win percentage: {c1_win_percentage:.2f}%")
print(f"{c2} win percentage: {c2_win_percentage:.2f}%")
